File: rust/src/189.轮转数组.rs
/*
 * @lc app=leetcode.cn id=189 lang=rust
 *
 * [189] 轮转数组
 */

// @lc code=start
impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        //使用三次翻转的方法，可以原地翻转
        let k = k as usize % nums.len();
        if k == 0 {
            return;
        }
        // rust的这个函数是用原地更新的办a实现翻转的
        nums.reverse();
        nums[..k].reverse();
        nums[k..].reverse();
    }
}
// @lc code=end

File: cpp/189.轮转数组.cpp
/*
 * @lc app=leetcode.cn id=189 lang=cpp
 *
 * [189] 轮转数组
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution {
public:
  void swap_part(vector<int> &nums, int l, int r) {
    while (l < r) {
      swap(nums[l++], nums[r--]);
    }
  }
  void rotate(vector<int> &nums, int k) {
    k %= nums.size();
    swap_part(nums, 0, nums.size() - 1 - k);
    swap_part(nums, nums.size() - k, nums.size() - 1);
    swap_part(nums, 0, nums.size() - 1);
  }
};
// @lc code=end
